move and revise AWS constraints (#846)
aws/karpenter: website/content/en/pre-docs/AWS/constraints.md

---

> title: "Provisioning Configuration"

Reviewer: Does it make more sense to call this page AWS Provider? And name the file AWS/provider.md

Author: I think "AWS Provider" is comperably accurate, however I want to give some indication as to what this page helps users do, which (from my perspective) is "configure the instances/nodes provisioned by the AWS provider", or "provisioning configuration"

linkTitle: "Provisioning"
weight: 10
---

> The [Provisioner CRD]({{< ref "../reference/provisioner-crd.md" >}}) provides two sections for constraining nodes. 

Reviewer: There's actually three (taints). I recommend avoiding framing things "n-ways" to do something unless there's something fundamental about why n will only ever be n. This is the type of detail that tends to not survive the test of time.

Author: only "spec.requirements" and "spec.provider" change the calls to EC2 right?

> - [`spec.requirements`](../reference/provisioner-crd#specrequirements)

Reviewer: Does it make sense to link to the NodeSelection section of the Tasks / Running Pods doc?

  - Cloud Provider Agnostic 
  - Kubernetes Well Known Labels
  - This section includes generally applicable constraints (zone, instance type) that each cloud provider is expected to implement. 
  - Reference the Provisioner CRD for more information. 
- [`spec.provider`](#specprovider)
  - Cloud Provider Specific
  - This section defines constraints that are unique to AWS, such as SecurityGroups.
  - Defined below. 

> ## spec.provider

Reviewer: I'd shortcircuit everything above and just talk about `provider` and provider specific things.

> Karpenter uses Kubernetes' `runtime.RawExtensions` to extend the [Provisioner CRD](../reference/provisioner-crd#s) with vendor specific configuration in [spec.provider].(https://github.com/awslabs/karpenter/blob/main/pkg/cloudprovider/aws/apis/v1alpha1/provider.go#L33). This section covers parameters of the AWS Cloud Provider.

Reviewer 2: I agree that we should cut the bullet list and just start in on the spec.provider content.

cloud provider specific constrains

[Review these fields in the code.](https://github.com/awslabs/karpenter/blob/main/pkg/cloudprovider/aws/apis/v1alpha1/provider.go#L33)

### InstanceProfile
An `InstanceProfile` is a way to pass a single IAM role to an EC2 instance.

It is required, and specified by name. A suitable `KarpenterNodeRole` is created in the getting started guide.

```
spec:
  provider:
    instanceProfile: MyInstanceProfile
```

### LaunchTemplate

A launch template is a set of configuration values sufficient for launching an EC2 instance (e.g., AMI, storage spec).

> A custom launch template be specified by name. If none is specified, Karpenter will automatically create a launch template.

Reviewer: ```suggestion
A custom launch template is specified by name. If none is specified, Karpenter will automatically create a launch template. Custom launch templates provide greater control of the underlying VM at the cost of complexity. You will need to manage security groups, kubelet versions, and kubelet parameters (e.g. labels, taints). 

```

Reviewer: Also -- can we link to the custom launch template guide here?

Review the [Launch Template documentation](launch-templates.md) to learn how to create a custom one.

```
spec:
  provider:
    launchTemplate: MyLaunchTemplate
```

### SubnetSelector

> By default, Karpenter discovers subnets by tags. Alternatively, cluster subnets may list specific subnets.

Reviewer: ```suggestion
Karpenter discovers subnets by tags.

```

> Subnets may be specified by AWS tag, or by name. Either approach supports wildcards. 

Reviewer: This is just tags. 
```suggestion
```

> When creating an instance, Karpenter picks a single subnet from this this. 

Reviewer: ```suggestion
When launching nodes, Karpenter automatically chooses a subnet that matches the desired zone. If multiple subnets exist for a zone, one is chosen randomly.

```

**Examples**

Select all subnets with a specified tag:
```
spec:
  provider:
    subnetSelector:
      kubernetes.io/cluster/MyCluster: '*'
```

> Select subnets by name:

Reviewer: ```suggestion
AWS implements subnet names using tags, which can explicitly selected
```

```
  subnetSelector:
    Name: subnet-0fcd7006b3754e95e
```

Select subnets by an arbitrary AWS tag key/value pair:
```
  subnetSelector:
    MySubnetTag: value

> Select subnets using wildcards:
> ```
>   subnetSelector:
>     Name: *public* 
>     MySubnetTag: '' # all resources with this tag
> 
> ```

Reviewer: This example is confusing to me. Let's scope it down to just wildcards. 
```suggestion
Subnets names often include whether or not they are exposed to the internet, you can select them using wildcards:
```
  subnetSelector:
    Name: *public* 
```

```

### SecurityGroupSelector

> Karpenter uses the EKS default security group, unless another is specified. The security group of an instance is comparable to a set of firewall rules.

Reviewer: This is inaccurate. Security group discovery works identically to subnet discovery
```suggestion
The security group of an instance is comparable to a set of firewall rules.
```

> EKS creates at least two security groups, [review the documentation](https://docs.aws.amazon.com/eks/latest/userguide/sec-group-reqs.html) for more info.

Reviewer: I'd remove this, since this is not EKS specific. If readers need to configure this, they will already know how/where they got their security groups.
```suggestion
```

Author: keeping with the cheat sheet concept, and the low cost of including a link, i'm going to keep it.

> Security Groups may be specified by AWS tag, or by name. Either approach supports wildcards. 

Reviewer: I'd mirror all of the comments above for security groups. They work exactly the same

> Each instance gets *all* of the listed security groups.

Reviewer: ```suggestion
When launching nodes, Karpenter uses all of the security groups that match the selector. The only exception to this is security groups tagged with the label `kubernets.io/cluster/${CLUSTER_NAME}`. The AWS Load Balancer controller requires that only a single security group with this tag may be attached to a node. In this case, Karpenter selects randomly.

```

**Examples**

Select all security groups with a specified tag:
```
spec:
  provider:
    securityGroupSelector:
      kubernetes.io/cluster/MyKarpenterSecurityGroups: '*'
```

Select security groups by name, or another tag:
```
 securityGroupSSelector:
   Name: sg-01077157b7cf4f5a8
   MySecurityTag: '' # matches all resources with the tag
```

Select security groups by name using a wildcard:
```
 subnetSelector:
   Name: *public*
```

### Tags

> All listed tags will be added to every node created by this provisioner.

Reviewer: ```suggestion
All listed tags will be added to every EC2 Instance launched by this provisioner.

```

```
spec:
  provider:
    tags:
      InternalAccountingTag: 1234
      dev.corp.net/app: Calculator
      dev.corp.net/team: MyTeam
```

> ## Other Resources

Reviewer: Good call to include this here. Custom Resources are definitely AWS specific.

---
title: "Provisioning Configuration"
linkTitle: "Provisioning"
weight: 10
---

The [Provisioner CRD]({{< ref "../reference/provisioner-crd.md" >}}) provides two sections for constraining nodes. 

- [`spec.requirements`](../reference/provisioner-crd#specrequirements)
  - Cloud Provider Agnostic 
  - Kubernetes Well Known Labels

>   - This section includues generally applicable constraints (zone, instance type) that each cloud provider is expected to implement. 

Reviewer 3: "includes" is misspelled. Are constraints such as zone, instance type, etc required or optional? Can we can get clarity from engineering?

  - Reference the Providioner CRD for more information. 
- [`spec.provider`](#specprovider)
  - Cloud Provider Specific
  - This section defines constraints that are unique to AWS, such as SecurityGroups.
  - Defined below. 

## spec.provider

cloud provider specific constrains

[Review these fields in the code.](https://github.com/awslabs/karpenter/blob/main/pkg/cloudprovider/aws/apis/v1alpha1/provider.go#L33)

### InstanceProfile

> An `InstanceProfile` is a region specific EC2 resource that is comprised of a reference to a single global IAM role.

Reviewer 3: I found this sentence to be a little confusing. Do you think we need to distinguish the scope of an InstanceProfile (regional) and an IAM role (global)? Can we say that an instance profile is a way to pass an IAM role to an EC2 instance?

Author: I personally think that distinction is noteworthy, but probabbly not in karpenter docs.

I revised and shortened.

It is required, and specified by name. A suitable `KarpenterNodeRole` is created in the getting started guide.

```
spec:
  provider:
    instanceProfile: MyInstanceProfile
```

### LaunchTemplate

> A launch template is a set of configuration values sufficent for launching an EC2 instance (e.g., AMI, storage spec).

Reviewer 3: "sufficient" is misspelled.

> A custom launch templay *may optionally* be specified by name. If none is specified, Karpenter will automatically create a launch template.

Reviewer 3: I think "templay" should be changed to "template". Not sure you need the word "optionally" but that's a stylistic thing.

Review the [Launch Template documentation](launch-templates.md) to learn how to create a custom one.

```
spec:
  provider:
    launchTemplate: MyLaunchTemplate
```

### SubnetSelector

> By default, Karpenter discovers subnets by tags. Alternatively, cluster subnets may be directly enumerated. 

Reviewer 3: I would probably say "Alternatively, you may list specific subnets." but that's a stylistic thing. Your choice.

Subnets may be specified by AWS tag, or by name. Either approach supports wildcards. 

> Each instance gets *one* subnet *chosen* from this list.

Reviewer 3: Each instance is deployed onto one, and only one, subnet from this list?

Author: revised.

**Examples**

Select all subnets with a specified tag:
```
spec:
  provider:
    subnetSelector:
      kubernetes.io/cluster/MyCluster: '*'
```

Select subnets by name:
```
  subnetSelector:
    Name: subnet-0fcd7006b3754e95e
```

Select subnets by an arbitary AWS tag key/value pair:
```
  subnetSelector:
    MySubnetTag: value

Select subnets using wildcards:
```
  subnetSelector:
    Name: *public* 
    MySubnetTag: '' # all resources with this tag

```

### SecurityGroupSelector

Karpenter uses the EKS default security group, unless another is specified. The security group of an instance is comperable to a set of firewall rules.

> EKS creates at least two security groups, review the documentation for more info.

Reviewer 3: Can you include a hyperlink to the documentation?

Author: good idea, done.

Security Groups may be specified by AWS tag, or by name. Either approach supports wildcards. 

Each instance gets *all* of the listed security groups.

**Examples**

Select all security groups with a specified tag:
```
spec:
  provider:
    securityGroupSelector:
      kubernetes.io/cluster/MyKarpenterSecurityGroups: '*'
```

Select security groups by name, or another tag:
```
 securityGroupSSelector:
   Name: sg-01077157b7cf4f5a8
   MySecurityTag: '' # matches all resources with the tag
```

Select security groups by name using a wildcard:
```
 subnetSelector:
   Name: *public*
```

### Tags

All listed tags will be added to every node created by this provisioner.

```
spec:
  provider:
    tags:
      InternalAccountingTag: 1234
      dev.corp.net/app: Calculator
      dev.corp.net/team: MyTeam
```

## Other Resources

### Accelerators, GPU

Accelerator (e.g., GPU) values include
- `nvidia.com/gpu`
- `amd.com/gpu`
- `aws.amazon.com/neuron`

Karpenter supports accelerators, such as GPUs.

To enable instances with accelerators, use the [instance type
well known label selector](#instance-types).

> Additionally, include a resource requirement in the workload manifest. Thus,

Reviewer 3: Can you replace "Thus" with "This will cause...to be scheduled onto the appropriate node"?

- [`spec.provider`](#specprovider)
  - Cloud Provider Specific
  - This section defines constraints that are unique to AWS, such as SecurityGroups.
  - Defined below. 

## spec.provider

This section covers parameters of the AWS Cloud Provider.

[Review these fields in the code.](https://github.com/awslabs/karpenter/blob/main/pkg/cloudprovider/aws/apis/v1alpha1/provider.go#L33)

### InstanceProfile
An `InstanceProfile` is a way to pass a single IAM role to an EC2 instance.

It is required, and specified by name. A suitable `KarpenterNodeRole` is created in the getting started guide.

```
spec:
  provider:
    instanceProfile: MyInstanceProfile
```

### LaunchTemplate

A launch template is a set of configuration values sufficient for launching an EC2 instance (e.g., AMI, storage spec).

A custom launch template is specified by name. If none is specified, Karpenter will automatically create a launch template.

Review the [Launch Template documentation](launch-templates.md) to learn how to create a custom one.

```
spec:
  provider:
    launchTemplate: MyLaunchTemplate
```

### SubnetSelector

> Karpenter discovers subnets using [AWS tags](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/Using_Tags.html). Alternatively, provisioners may list specific subnets.

Reviewer: This is a bit inaccurate. The way you specify specific subnets is via tag discovery. 
```suggestion
Karpenter discovers subnets using [AWS tags](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/Using_Tags.html). 
```

> Subnets may be specified by any AWS tag, including "name". Selecting tags using wildcards ("*") is supported.

Reviewer: ```suggestion
Subnets may be specified by any AWS tag, including `Name`. Selecting tag values using wildcards ("*") is supported.

```

When launching nodes, Karpenter automatically chooses a subnet that matches the desired zone. If multiple subnets exist for a zone, one is chosen randomly.

**Examples**

Select all subnets with a specified tag:
```
spec:
  provider:
    subnetSelector:
      kubernetes.io/cluster/MyCluster: '*'
```

Select subnets by name:
```
  subnetSelector:
    Name: subnet-0fcd7006b3754e95e
```

Select subnets by an arbitrary AWS tag key/value pair:
```

>   subnetSelector:

Reviewer: I'd either include spec.provider in all of these examples, or none of them.

    MySubnetTag: value

Select subnets using wildcards:
```
  subnetSelector:
    Name: *public* 

```

### SecurityGroupSelector

The security group of an instance is comparable to a set of firewall rules.
If no security groups are explicitly listed, Karpenter discovers them using the tag "kubernetes.io/cluster/MyClusterName", similar to subnet discovery.

EKS creates at least two security groups by default, [review the documentation](https://docs.aws.amazon.com/eks/latest/userguide/sec-group-reqs.html) for more info.

Security groups may be specified by any AWS tag, including "name". Selecting tags using wildcards ("*") is supported.

‼️ When launching nodes, Karpenter uses all of the security groups that match the selector. The only exception to this is security groups tagged with the label `kubernets.io/cluster/MyClusterName`. The AWS Load Balancer controller requires that *only a single security group with this tag may be attached to a node*. In this case, Karpenter selects randomly.

**Examples**

Select all security groups with a specified tag:
```
spec:
  provider:
    securityGroupSelector:
      kubernetes.io/cluster/MyKarpenterSecurityGroups: '*'
```

> Select security groups by name, or another tag:

Reviewer: Seems a bit odd that this section doesn't have parity with the subnets section.

```
 securityGroupSSelector:
   Name: sg-01077157b7cf4f5a8
   MySecurityTag: '' # matches all resources with the tag
```

Select security groups by name using a wildcard:
```
 subnetSelector:
   Name: *public*
```

### Tags

Tags will be added to every EC2 Instance launched by this provisioner.

```
spec:
  provider:
    tags:
      InternalAccountingTag: 1234
      dev.corp.net/app: Calculator
      dev.corp.net/team: MyTeam
```

## Other Resources

### Accelerators, GPU

Accelerator (e.g., GPU) values include
- `nvidia.com/gpu`
- `amd.com/gpu`
- `aws.amazon.com/neuron`

Karpenter supports accelerators, such as GPUs.

> To enable instances with accelerators, use the [instance type
> well known label selector](#instance-types).

Reviewer: This is inaccurate. Karpenter will automatically filter down instance types if you specify gpus in your resources. 
```suggestion
```
